src/main/kotlin/com/openlattice/data/storage/IndexingMetadataManager.kt

```diff
@@ -71,20 +64,16 @@ class IndexingMetadataManager(private val hds: HikariDataSource, private val par
                 .getPartitionsByEntitySetId(linkingIdsWithLastWrite.keys)
                 .mapValues { it.value.toList() }
 
-        return linkingIdsWithLastWrite.map { (entitySetId, entities) ->
-            val partitions = entitySetPartitions.getValue(entitySetId)
-            entities.entries
-                    .groupBy({ getPartition(it.key, partitions) }, { it.value })
-                    .toSortedMap()
-                    .map { (partition, idsAndExpiration) ->
-
-                        lockIdsAndExecuteAndCommit(
-                                hds,
-                                updateLastLinkingIndexSql,
-                                entitySetId,
-                                partition,
-                                shouldLockEntireEntitySet = true
-                        ) { ps ->
+        return hds.connection.use { connection ->
+            val ps = connection.prepareStatement(updateLastLinkingIndexSql)
+            linkingIdsWithLastWrite.map { (entitySetId, entities) ->
+                val partitions = entitySetPartitions.getValue(entitySetId)
+                entities.entries
```

Comment:
Can we pull the grouping/sorting logic outside of the connection? (Just to reduce the amount of work that needs to be done while it's been reserved)

Reply:
That's what it was before. I pulled it out to avoid re-creating the prepared statement over and over in the loop. I could do something like building out all the groupings before executing the queries, but that's a bigger change.

Reply:
can't we just assign the sorted map to a variable before acquiring the connection?

Reply:
(then acquire the connection, prepare the statement, iterate over the sorted map)

Reply:
Yeah that would work, but you still have this: https://jdbc.postgresql.org/documentation/head/server-prepare.html

So to get the full benefit of both you have to prepare the full data structure ahead of time.

Reply:
LATTICE-2530